compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeModifierManager.java

```diff
@@ -0,0 +1,437 @@
+                    Objects.equals(compilerPluginInfo, that.compilerPluginInfo);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(codeModifier, compilerPluginInfo);
+        }
+    }
+
+    private static class CodeModifyTaskResult {
+        private final List<Diagnostic> reportedDiagnostics;
+        private final Map<DocumentId, ModifiedSourceFile> sourceFilesMap;
+
+        public CodeModifyTaskResult(List<Diagnostic> reportedDiagnostics,
+                                    Map<DocumentId, ModifiedSourceFile> sourceFilesMap) {
+            this.reportedDiagnostics = reportedDiagnostics;
+            this.sourceFilesMap = sourceFilesMap;
+        }
+
+        Collection<Diagnostic> reportedDiagnostics() {
+            return reportedDiagnostics;
+        }
+
+        ModifiedSourceFile sourceFile(DocumentId documentId) {
+            return sourceFilesMap.get(documentId);
+        }
+
+        Collection<DocumentId> documentIds() {
+            return sourceFilesMap.keySet();
+        }
+
+        boolean containsSourceFile() {
+            return !sourceFilesMap.isEmpty();
+        }
+    }
+
+    private static class CodeModifierTaskResultBuilder {
+        private final List<Diagnostic> reportedDiagnostics;
+        private final List<ModifiedSourceFile> modifiedSourceFiles;
+
+        CodeModifierTaskResultBuilder() {
+            this.reportedDiagnostics = new ArrayList<>();
+            this.modifiedSourceFiles = new ArrayList<>();
+        }
+
+        CodeModifierTaskResultBuilder addDiagnostics(Collection<Diagnostic> diagnostics) {
+            reportedDiagnostics.addAll(diagnostics);
+            return this;
+        }
+
+        CodeModifierTaskResultBuilder addSourceFiles(Collection<ModifiedSourceFile> sourceFiles) {
+            modifiedSourceFiles.addAll(sourceFiles);
+            return this;
+        }
+
+        CodeModifyTaskResult build() {
+            Map<DocumentId, ModifiedSourceFile> sourceFilesMap = new HashMap<>();
+            for (ModifiedSourceFile sourceFile : modifiedSourceFiles) {
+                sourceFilesMap.put(sourceFile.documentId(), sourceFile);
+            }
+            return new CodeModifyTaskResult(reportedDiagnostics, sourceFilesMap);
+        }
+    }
+
+    private static class PackageModifier {
+
+        Package modifyPackage(Package currentPackage, CodeModifyTaskResult codeModifyTaskResult) {
+            Package newPackage = currentPackage;
+            for (DocumentId documentId : codeModifyTaskResult.documentIds()) {
+                newPackage = modifyModule(documentId.moduleId(), newPackage, codeModifyTaskResult);
+            }
+            return newPackage;
+        }
+
+        private Package modifyModule(ModuleId moduleId, Package pkg, CodeModifyTaskResult codeModifyTaskResult) {
+            Module module = pkg.module(moduleId);
+            Module.Modifier modifier = module.modify();
```

Comment:
calling module.modify() will return a new Modifier instance everytime. We would loose modifications if more than one document is modified i na module

Reply:
Changed related test case to have multiple bal files in this [commit](5623cfa750d66a06a47f7fb97795397e056b4229), and it's working as expected.